### DISCUSSION TITLE
Rest API for Different schemas

### DISCUSSION THREAD

**Discussion by erickscarponeti** [supabase/supabase#3875] [Questions]
We can do it now? or just the public schema generate rest apis? Someone can give me a help?

**Comment by dshukertjr** [comment]
<strike>Not at the moment, but it seems like Supabase team is very close to implementing this. 

You can follow [this](https://github.com/supabase/supabase/discussions/1222) discussion for updates on accessing non-public schema from Supabase. </strike>

edited: view post from @bnjmnt4n

**Comment by bnjmnt4n** [comment]
Actually, this feature is now available on the platform: see https://supabase.io/docs/reference/javascript/initializing#api-schemas.

**Reply by dshukertjr** [reply]
That's amazing!

**Comment by erickscarponeti** [comment]
But doesnt generate restful api for all, just expose for other languages

**Reply by MildTomato** [reply]
You can choose which schemas to expose in the settings page.
As @bnjmnt4n's [answer](https://github.com/supabase/supabase/discussions/3875#discussioncomment-1631066) link says.

"Go to Settings > API > Schema and add the schema which you want to expose to on the API."

**Reply by erickscarponeti** [reply]
<img width="1368" alt="Captura de Tela 2021-11-16 às 15 11 44" src="https://user-images.githubusercontent.com/72212010/142041785-d8cc91ba-0138-46f9-b22c-6830f45853cb.png">

Do you know how can manipulate with these URL's? Because expose in settings, still show only the public schema urls to make calls

**Reply by leo-fe99** [reply]
would like to know to :smile:

**Reply by silentworks** [reply]
The schema doesn't get added to the URL, its added as a header called `accept-profile`. You just need to set the table name as these examples show, so if you have a `users` table in a schema called `hello` the URL would still be `https://project_ref.supabase.co/rest/v1/users?select=*` with a header of `accept-profile=hello`. 

Do note that new schema and tables need to have the correct permissions set in order to access them thru the client library or the endpoint.

EDIT: You will have to grant usage to the schema using
```sql
GRANT USAGE TO SCHEMA schema_name TO role_name;
```
In my case the schema_name is `hello` and the roles I want to have access are `anon, authenticated`.
```sql
GRANT USAGE TO SCHEMA hello TO anon, authenticated;
```
For the table itself you will also need to grant the roles access.
```sql
GRANT ALL ON table_name IN SCHEMA schema_name TO role_name
``` 
In this case my table_name is `users`, schema_name is `hello` and the roles I want to have access are `anon, authenticated`.
```sql
GRANT ALL ON "users" IN SCHEMA "hello" TO anon, authenticated;
```

**Reply by vegaboo** [reply]
Would the same logic work for `graphql` api?

**Reply by tonyxiao** [reply]
see https://github.com/supabase/supabase/discussions/7957
